File: peri.dot/version/exceptions.py
from .modules.colorama import init, Fore, Style
init()

class Exc_Error():
    def __init__(self, exc, msg, start, end, context):
        self.exc = exc
        self.msg = msg
        self.start = start
        self.end = end
        self.context = context

    def asstring(self):
        result = self.traceback()
        result += f'''   >{Fore.YELLOW}{self.start.lntext}{Style.RESET_ALL}
    {Fore.YELLOW}{' ' * self.start.column}{'^' * (self.end.column - self.start.column)}{Style.RESET_ALL}
{Fore.RED}{Style.BRIGHT}{self.exc}{Style.RESET_ALL}: {Fore.RED}{self.msg}{Style.RESET_ALL}'''
        return(result)

    def traceback(self):
        result = ''
        pos = self.start
        context = self.context

        while context:
            result = f'''  {Fore.GREEN}File {Style.BRIGHT}{pos.file}{Style.RESET_ALL}, {Fore.GREEN}Line {Style.BRIGHT}{pos.line}{Style.RESET_ALL}{Fore.GREEN}, Column {Style.BRIGHT}{pos.column}{Style.RESET_ALL}{Fore.GREEN}, In {Style.BRIGHT}{context.display}{Style.RESET_ALL}\n''' + result

            pos = context.parententrypos
            context = context.parent

        result = f'{Fore.BLUE}{Style.BRIGHT}Traceback (most recent call last):{Style.RESET_ALL}\n' + result
        return(result)



class Exc_SyntaxError(Exc_Error):
    def __init__(self, msg, start, end, context=None):
        super().__init__('SyntaxError', msg, start, end, context)

class Exc_EscapeError(Exc_Error):
    def __init__(self, msg, start, end, context=None):
        super().__init__('EscapeError', msg, start, end, context)



class Cmd_CmdError():
    def __init__(self, exc, msg, argnum, arg):
        self.exc = exc
        self.msg = msg
        self.argnum = argnum
        self.arg = arg

    def asstring(self):
        result = f'''{Fore.BLUE}{Style.BRIGHT}An error occured while reading arguments{Style.RESET_ALL}
  {Fore.GREEN}Argument{Style.RESET_ALL} {Fore.GREEN}{Style.BRIGHT}{self.argnum}{Style.RESET_ALL}
   >{Fore.YELLOW}{self.arg}{Style.RESET_ALL}
    {Fore.YELLOW}{'^' * len(self.arg)}{Style.RESET_ALL}
{Fore.RED}{Style.BRIGHT}{self.exc}{Style.RESET_ALL}: {Fore.RED}{self.msg}{Style.RESET_ALL}'''
        return(result)

class Cmd_CmdArgumentError(Cmd_CmdError):
    def __init__(self, msg, argnum, arg):
        super().__init__('CmdArgumentError', msg, argnum, arg)

class Cmd_NotSupportedError(Cmd_CmdError):
    def __init__(self, msg, argnum, arg):
        super().__init__('NotSupportedError', msg, argnum, arg)

File: src/version/exceptions.rs
use std::fmt;
use colored::*;
use md5::{Md5, Digest};

use super::lexer;
use super::context;



pub struct CommandLineException {
    pub name     : String,
    pub msg      : String
}
impl fmt::Display for CommandLineException {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}: {}", self.name.red().bold(), self.msg.red())
    }
}



pub struct LexerException {
    pub failed: bool,
    pub name  : String,
    pub msg   : String,
    pub ucmsg : String,
    pub start : lexer::LexerPosition,
    pub end   : lexer::LexerPosition
}
impl fmt::Display for LexerException {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut exc = "".to_string();
        exc += format!("{}\n", "An exception occured while generating tokens:".blue().bold()).as_str();
        exc += format!("  {} `{}`,\n", "File".green(), self.start.file.green().bold()).as_str();
        exc += format!("  {} {}, {} {}\n", "Line".green(), (self.start.line + 1).to_string().green().bold(), "Column".green(), (self.start.column + 1).to_string().green().bold()).as_str();
        exc += format!("    {}\n", self.start.lines[self.start.line].yellow().bold()).as_str();
        exc += format!("    {}{}\n", " ".repeat(self.start.column), "^".repeat(self.end.index - self.start.index).yellow()).as_str();
        let mut errorcode = Md5::new();
        errorcode.input(format!("{}{}", self.name, self.ucmsg));
        let errorcode = format!("{:?}", errorcode.result()).replace(", ", "").replace("[", "").replace("]", "");
        let errorcode = errorcode.get(0..5).unwrap();

        exc += format!("[{}{}] {}: {}", "e".red(), errorcode.red().bold(), self.name.red().bold(), self.msg.red()).as_str();

        write!(f, "{}", exc)
    }
}



#[derive(Clone)]
pub struct ParserException {
    pub failed: bool,
    pub name  : String,
    pub msg   : String,
    pub ucmsg : String,
    pub start : lexer::LexerPosition,
    pub end   : lexer::LexerPosition
}
impl fmt::Display for ParserException {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut exc = "".to_string();
        exc += format!("{}\n", "An exception occured while parsing tokens:".blue().bold()).as_str();
        exc += format!("  {} `{}`,\n", "File".green(), self.start.file.green().bold()).as_str();
        exc += format!("  {} {}, {} {}\n", "Line".green(), (self.start.line + 1).to_string().green().bold(), "Column".green(), (self.start.column + 1).to_string().green().bold()).as_str();
        exc += format!("    {}\n", self.start.lines[self.start.line].yellow().bold()).as_str();
        exc += format!("    {}{}\n", " ".repeat(self.start.column), "^".repeat(self.end.index - self.start.index).yellow()).as_str();
        let mut errorcode = Md5::new();
        errorcode.input(format!("{}{}", self.name, self.ucmsg));
        let errorcode = format!("{:?}", errorcode.result()).replace(", ", "").replace("[", "").replace("]", "");
        let errorcode = errorcode.get(0..5).unwrap();

        exc += format!("[{}{}] {}: {}", "e".red(), errorcode.red().bold(), self.name.red().bold(), self.msg.red()).as_str();

        write!(f, "{}", exc)
    }
}



#[derive(Clone)]
pub struct InterpreterException {
    pub failed : bool,
    pub name   : String,
    pub msg    : String,
    pub ucmsg  : String,
    pub start  : lexer::LexerPosition,
    pub end    : lexer::LexerPosition,
    pub context: Option<context::Context>
}
impl InterpreterException {
    fn generatetraceback(self) -> String {
        let mut result = "".to_string();
        let mut part: String;
        let mut pos = self.start;
        let mut ctx = self.context;
        let mut context: context::Context;

        loop {
            match ctx {
                Some(value) => {context = value}
                None        => break
            }

            part = format!("  {} `{}`, {} `{}`,\n", "File".green(), pos.file.green().bold(), "In".green(), context.display.green().bold());
            part += format!("  {} {}, {} {}\n", "Line".green(), pos.line.to_string().green().bold(), "Column".green(), pos.column.to_string().green().bold()).as_str();
            result = part + result.as_str();

            match context.parententry {
                Some(value) => {pos = value}
                None        => break
            }
            ctx = *context.parent;
        }

        result = format!("{}\n{}", "Traceback (Most recent call last):".blue().bold(), result.as_str());
        return result;
    }
}
impl fmt::Display for InterpreterException {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut exc = "".to_string();
        exc += self.clone().generatetraceback().as_str();
        exc += format!("    {}\n", self.start.lines[self.start.line].yellow().bold()).as_str();
        exc += format!("    {}{}\n", " ".repeat(self.start.column), "^".repeat(self.end.index - self.start.index).yellow()).as_str();
        let mut errorcode = Md5::new();
        errorcode.input(format!("{}{}", self.name, self.ucmsg));
        let errorcode = format!("{:?}", errorcode.result()).replace(", ", "").replace("[", "").replace("]", "");
        let errorcode = errorcode.get(0..5).unwrap();

        exc += format!("[{}{}] {}: {}", "e".red(), errorcode.red().bold(), self.name.red().bold(), self.msg.red()).as_str();

        write!(f, "{}", exc)
    }
}
